ably/realtime_channel_spec_test.go

```diff
@@ -155,6 +155,662 @@ func attachAndWait(t *testing.T, channel *ably.RealtimeChannel) {
+		}
+		ablytest.Instantly.NoRecv(t, &change, stateChanges, t.Fatalf)
+	})
+
+	t.Run("RTL5g", func(t *testing.T) {
+		in, _, c, channel, stateChanges, _ := setup(t)
+		channel.OnAll(stateChanges.Receive)
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		connectionChange := make(chan ably.ConnectionStateChange)
+		c.Connection.OnAll(func(change ably.ConnectionStateChange) {
+			connectionChange <- change
+		})
+
+		// Get the channel to ATTACHED.
+
+		in <- &proto.ProtocolMessage{
+			Action:  proto.ActionAttached,
+			Channel: channel.Name,
+		}
+
+		var change ably.ChannelStateChange
+
+		ablytest.Soon.Recv(t, &change, stateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateAttached, change.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, change)
+		}
+
+		c.Close()
+		// set connection state to closing
+		c.Connection.SetState(ably.ConnectionStateClosing, nil, time.Minute)
+
+		ablytest.Instantly.Recv(t, nil, connectionChange, t.Fatalf) // Consume connection state change to closing
+		err := channel.Detach(ctx)
+		if expected, got := "cannot Detach channel because connection is in CLOSING state", err.Error(); !strings.Contains(got, expected) {
+			t.Fatalf("expected error %+v; got %v", expected, got)
+		}
+
+		// set connection state to failed
+		c.Connection.SetState(ably.ConnectionStateFailed, nil, time.Minute)
+		ablytest.Instantly.Recv(t, nil, connectionChange, t.Fatalf) // Consume connection state change to closing
+		err = channel.Detach(ctx)
+		if expected, got := "cannot Detach channel because connection is in FAILED state", err.Error(); !strings.Contains(got, expected) {
+			t.Fatalf("expected error %+v; got %v", expected, got)
+		}
+	})
+
+	t.Run("RTL5h : Connection state initialized", func(t *testing.T) {
+		in, out, c, channel, _, _ := setup(t)
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		connectionChange := make(chan ably.ConnectionStateChange)
+		c.Connection.OnAll(func(change ably.ConnectionStateChange) {
+			connectionChange <- change
+		})
+
+		channelStateChange := make(chan ably.ChannelStateChange)
+		channel.OnAll(func(change ably.ChannelStateChange) {
+			channelStateChange <- change
+		})
+		var outMsg *proto.ProtocolMessage
+
+		// Get the channel to ATTACHED.
+		in <- &proto.ProtocolMessage{
+			Action:  proto.ActionAttached,
+			Channel: channel.Name,
+		}
+
+		var channelChange ably.ChannelStateChange
+
+		ablytest.Instantly.Recv(t, &channelChange, channelStateChange, t.Fatalf) // Consume attached message
+
+		if expected, got := ably.ChannelStateAttached, channelChange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelChange)
+		}
+
+		var change ably.ConnectionStateChange
+
+		// set connection state to ConnectionStateInitialized
+		c.Connection.SetState(ably.ConnectionStateInitialized, nil, time.Minute)
```

Comment:
Same point as here https://github.com/ably/ably-go/pull/299#discussion_r599334068

This is especially trivial to avoid, since INITIALIZED is the connection's initial state when autoConnect=false.

Reply:
I am not sure how to test this, because it's difficult to reproduce the connection state to Initialized once the connection
is established with the server. In order to run the test for detach, it needs to be attached first, hence connected in the first place. You cannot go back to initialized once connected. Maybe, we should go back to the point we raised about attach sometime back.

Reply:
https://ably-real-time.slack.com/archives/C0114KK7S6R/p1617842973002400

Reply:
We should remove initialized condition from the spec I feel, there's no point in having it there ... since practically it's impossible to reach the state at least for detach operation

Reply:
> In order to run the test for detach, it needs to be attached first

Not in this case. This spec item covers what happens if you call `detach` without having attached first.

That said, as discussed in private, I agree it would make more sense if the operation failed immediately.

Anyway, for the purposes of this test, this state can be reached just with:

```go
r, _ := ably.NewRealtime(... WithAutoConnect(false) ...)
c := r.Channels.Get("foo")
c.Detach(ctx)
```

Reply:
Okay, makes sense : ), thanks for the explaination

Reply:
Channel state would be initialized without calling attach first, if that's the case, spec ([RTL5a](https://docs.ably.io/client-lib-development-guide/features/#RTL5a)) already covers it and it's not about queuing the message but doing nothing. I believe we already have a test for the same.

Reply:
Again coming back to the same point, maybe it should be removed from the spec same as `attach`

Reply:
That's true, unless I'm missing something there's no way for the connection to be INITIALIZED while the channel isn't INITIALIZED, so RTL5a covers this case already.

This part of the test can be removed then (CONNECTING and DISCONNECTED still apply, I think).

Maybe you want to raise this in ably/docs?

Reply:
Yeah, I will raise the PR in the docs